Attribute related fixes

cstamas opened the pull request:
This pull request contains following substantial fixes:
- (F1) TransitioningAttributeStorage made less eager. Fix made attribute-related IO ops from "read, read, write, delete" to "read, read" (worst cases).
- (F2) RecreateAttributesWalker operated on groups too, hence, group member repositories was walked for upgrade multiple times.
- (F3) RecreateAttributesWalker was forcing re-checksumming.
- (F4) on upgrade, attribute save happens only when it originates from "legacy" store (ie. partially upgraded repository will perform saves only on non-upgraded attributes).

And some other minor improvements regarding overall IO load of upgrade.

This pull request contains following major improvements:
- Nexus Core Walker component supports walk-throttling, it is possible to "interfere" with walking speed from now on (by slowing it down, not by walking faster than your HW is capable of! ;)
- Two out-of-the-box throttle controllers implemented: "fixed rate" and "adaptive" (the 2nd is TBD, not yet implemented!).
- AttributeUpgrader component exposed over JMX, and properties like maxTPS are even settable live during upgrade. Also, the upgrade process itself can be invoked/started over JMX (ie. you defer upgrade auto-start by system property to be able to set up for proper JMX monitoring).
- NumberSequence got new member: "linear" number sequence, and it's API is enhanced (got prev() method for stepping back).

The "initial" maxTPS value is set to 100 upgrade/sec, which would upgrade Central repository hosted in Nexus in 5 hours. But, again, the maxTPS is settable over JMX.

cstamas committed:
Making TransitioningAttributeStorage "non-eager".

Attribute upgrade from now on happens only on attribute STORE operation,
while attribute READ operation just _reads_ them (from new or falls back to old
storage).

cstamas committed:
Implemented Walk throttling.

Walker component got "throttling" support in very non-intrusive way.
Everything works as before, but there is now possibility to implement
real "throttling" too, ie. intentionally slowing down the walks if needed.

This is currently used in AttributeUpgrade background thread only, as we detected
that thread priority has nothing to do with IO (but only with CPU) and that
presence of this thread might still saturate IO of a server (and leaving CPU
under utilized, hence, JVM does all "good" with this low prio thread).

But with this change, every "walk" is now capable of being throttled even without
(or with slight) code change.

The current implementation should be used as "last resort"! If we need this at all,
using the system property override (to set "throttle time in millis") could be used
to empirically "tune" and find best values for given (hence, similar) Nexus instance.

By default, upgrade thread is NOT throttled.

cstamas committed:
Improvements to throttle support, and more.

Throttle support has a major improvements, now we can control throughput in
"transaction per second" manner too. Heavily using it in very improved UpgraderThread
that dynamically adapts itself to the limit "TPS".

cstamas committed:
Smaller style and other fixes.

Added missing source headers, javadoc'd and applied some comments.

cstamas committed:
Fixes for recreate attributes.

It was doing way too much.
Just as example: on my Mac, max UPS before this change was around 50, now it's over 250.

cstamas committed:
Improved logging.

cstamas committed:
Set "default" limit to 200 up/sec.

cstamas committed:
Exposing upgrade throttle controls and invocation controls over JMX.

cstamas committed:
Lowering default upgrade TPS to 100.

cstamas committed:
Extending NumberSequence API and introducing a new one.

Aside of "constant" and "fib" number sequences, there is now
"linear" number sequence, that is modeling a linear equation.
The number sequences in general got another important role in Nexus
(aside of current, to maintain auto blocking times of proxy repositories)
they are now controlling the "steepness" (ie. force expressed to brake pedal)
in case of throttle controllers.

cstamas committed:
NumberSequence UTs

cstamas committed:
Refactored throttle controllers to make them reusable.

As Walker component already supports this, making controllers reusable make them
usable in this very moment across Nexus.

cstamas committed:
No actual change, just variable renames to better reflect their meaning.

cstamas committed:
Improving precision in calculations and slight API change
for better stats.

cstamas committed:
Added UT for only one implemented throttle controller.

cstamas committed:
Added missing source headers.

cstamas committed:
IO improvement.

Only "legacy" attributes (those coming from legacy store) are saved. They are marked by
TransitioningAttributeStorage that is used ONLY when Nexus detects presence of old attribute
folder.

cstamas committed:
NEXUS-4770: Fixed robustness of upgrade.

Walker was already fixed (the original stack traces reported in issue).
Now the thread properly handle other exceptional cases like unexpected
problem during upgrade, the repo will be skipped. Also, repositories put
out of service. Before, even those two events above would cause upgrade
to report "all done", while it was actually not done.

As for retries, upgrade will be tried again on next Nexus boot, or it can
be invoked over JMX.

cstamas commented:
CI put it's +1 (feature IT build):
https://builds.sonatype.org/view/nexus/job/nexus-oss-its-feature/181/

The reason of build failure was actually missing headers from newly added sources, fixed in 9e0c13f439b3996372f5c3b69f73303a26c48709. Headers are checked after all the tests were passing.

mah0526 commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/FixedRateWalkerThrottleController.java:
```diff
@@ -0,0 +1,277 @@
+            if ( lastSliceTps > limiterTps )
+            {
+                // hold down the horses, increase sleepTime
+                if ( currentSleepTime.peek() <= 0 )
```
If I'm looking at this correctly, with a starting delay (NumberSequence) of 0, it seems you will never slow down -- i.e. you will never end up calling currentSleepTime.next().  If so, we should probably handle differently.

cstamas replied:
:beer: X1

cstamas replied:
Fixed in subsequent commit.

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/AttributeUpgraderThread.java:
```diff
@@ -0,0 +1,171 @@
+            }
+            else
+            {
+                logger.info( "Legacy attribute directory upgrade PARTIALLY finished. Please fix the problems and retry the upgrade." );
```
In the case of a problem with the upgrade where a WalkerException was thrown, that may be the case but the note about "fix the problems" should probably go in that log message (keeping the pertinent information for the user in one place), and perhaps logged as at ERROR (versus WARN).  Additionally, for upgrades on repos out of service that are not performed, it might be best just to tell the user that it will automatically be done at a later time (currently, on server restart I assume).  With those changes, then this message merely needs to note that the "...upgrade was partially completed.  Please see prior log messages for details."  There may not actually be any problems, and if there are, we should indicate to the user where they should start looking.  Of course, most users will probably never see this stuff anyway.

cstamas replied:
Fixed in subsequent commit.

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/AttributeUpgraderThread.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    public void onAdjustment( final FixedRateWalkerThrottleController controller )
+    {
+        logger.info(
```
Probably better to have this at DEBUG level for final release.

cstamas replied:
Fixed in subsequent commit.

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    @Override
+    public synchronized void upgradeAttributes( final boolean force )
```
What's the least amount of locking required for correctness?  Will this method acquire the dreaded configuration lock through the getLegacyAttributesDirectory() call?  If so and you need sync on entire method, you might want to make the external calls first and then lock.

cstamas replied:
No, this is the least locking needed, to retain correctness only. This method is synchronized _only_ to ensure that it will not spawn multiple upgrade thread over each other.

Also, if you look at the method body, you will notice that the time spent here is minimal (best case logging something and getting out), worst case creating a thread and starting it.

This synchronization is needed because the method can be invoked over JMX. Actually, the "execution path" that brings you here to this method is either the NexusStartedEvent (guaranteed to happen only once, and it's not the reason of syncing), _or_ it's been fired over JMX. And the JMX path is the one needing sync (2nd incoming call will simply "bounce off", if the thread is be alive and running).

mah0526 replied:
You can't make the calls that call out of this class (i.e. into config) before acquiring the lock?  I'll submit that logging calls are not an issue, but even if this is not a problem with the current implementation, it leaves a potential lingering issue for someone (with potentially less knowledge) to contend with down the road.

cstamas replied:
This class is about attribute upgrade. It does not modify config. But it's possible I am missing your point...

mah0526 replied:
No problem.  For example, the calls to getLegacyAttributesDirectory() then call applicationConfiguration.getWorkingDirectory(), diving into the internals of Nexus configuration.

cstamas replied:
Bad named class: almost all applicationConfiguration's methods are about "meta configuration" actually -- about various folders and their whereabout on disk, like nexus/conf folder is (this statement is a bit shaded with the presence of method saveConfiguration()). 

What you think is probably NexusConfiguration (that do extend ApplicationConfiguration)... that one DO expose things that could _modify_ configuration.

Again, getWorkingDir() method (or actually any other code within body of the synchronized method) does _not_ change configuration (nor does the spawned thread either).

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,296 @@
+        }
+        catch ( IOException e )
+        {
+            // hum?
```
Seems like we ought to log something here, as unlikely as the case may be.

cstamas replied:
Fixed in subsequent commit.

cstamas committed:
Fixing reported remarks.

cstamas committed:
Fixes for fixed rate throttle controller.

* NumberSequence is heavily checked for compliance with algorithm (some sequence might cause never to throttle, ie. those supplying negative numbers after being reset)
* Now it is possible to supply number sequence that has initial value of any non-negative value (before only strictly positives was possible)
* Increased "range" of throttling, value 0 means literally "I want to be paused/sleep" now (before it was do not throttle me)
* Improved Javadocs

cstamas committed:
Hit it hard, then continue with purring.

Thanks to enhancements, not it's possible to have fib sequence start with 0.

0,5,5,10,15,25,40... is how this will be slowed down.

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/walker/DefaultThrottleInfo.java:
```diff
@@ -0,0 +1,74 @@
+ * @author cstamas
+ * @since 2.0
+ */
+public class DefaultThrottleInfo
```
No use for a lastProcessTimeSpent() call?  Not sure if it is being calculated externally.

cstamas replied:
No use, but it might get used in near future, or we can easily drop it if it proves un-useful (after this is merged but before release)

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,295 @@
+ * rebuild attributes background task if needed. Finally, it leaves "marker" file to mark the fact upgrade did happen,
+ * to not kick in on any subsequent reboot.
+ * 
+ * @since 1.10.0
```
wrong since

cstamas replied:
Fixed in subsequent commit.

velo commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/AttributeUpgraderThread.java:
```diff
@@ -0,0 +1,205 @@
+import org.sonatype.nexus.proxy.walker.FixedRateWalkerThrottleController.FixedRateWalkerThrottleControllerCallback;
+import org.sonatype.nexus.util.FibonacciNumberSequence;
+
+public class AttributeUpgraderThread
```
JDoc missing here

cstamas replied:
Fixed in subsequent commit.

velo commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgraderMBean.java:
```diff
@@ -0,0 +1,82 @@
+
+import javax.management.StandardMBean;
+
+public class DefaultAttributeUpgraderMBean
```
JDoc

cstamas replied:
Fixed in subsequent commit.

mah0526 commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/walker/DefaultWalker.java:
```diff
@@ -315,6 +323,18 @@ protected void processItem( WalkerContext context, StorageItem item )
+            {
+                final long throttleTime = context.getThrottleController().throttleTime( info );
+
+                if ( throttleTime > 0 )
```
Only a thought... Worthwhile to encapsulate the Thread.sleep() call within the throttle controller classes?  This class probably need not care about those details (like having to get the time back and deal with it).  Perhaps refactor throttleTime( info ) to throttle( info ) and handle everything at a lower level of abstraction.

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,295 @@
+            {
+                // nag the user to remove the directory
+                getLogger().info(
+                    "Legacy attribute directory present, but is marked already as upgraded. Please delete, move or rename the \"{}\" folder.",
```
folder -> directory 

... use one term or the other

cstamas replied:
Fixed in subsequent commit.

mah0526 commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/FibonacciNumberSequence.java:
```diff
@@ -36,15 +38,20 @@ public FibonacciNumberSequence()
         reset();
     }
 
     public long next()
```
This logic looks flawed (though not introduced by this commit).  A call to next() should (I think) return the next number in the sequence.  However, it returns the current value in the sequence and advances "a" to the next number in the sequence.  Only a subsequent call to peek() will get the real next value.

mah0526 commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/FibonacciNumberSequence.java:
```diff
@@ -23,7 +23,9 @@
     implements NumberSequence
 {
-    private final long start;
+    private final long startA;
```
How about refactoring "startA" to "currentValue" and "startB" to "precedingValue"?  In the constructor, startA and startB could be replaced with startValue, precedingValue.  That would probably add a lot of clarity.

mah0526 commented on nexus/nexus-utils/src/main/java/org/sonatype/nexus/util/LinearNumberSequence.java:
```diff
@@ -0,0 +1,66 @@
+ * 
+ * @author cstamas
+ */
+public class LinearNumberSequence
```
Wait, this is a discontinuous, non-linear function :)

cstamas replied:
you mean "discrete function"? ;)

bdemers commented:
should add a test for broken attributes durring upgrade, and the case you brought up about three consecutive IOExceptions

cstamas committed:
Smaller fixes, javadocing.

cstamas commented:
By merging (and hence, closing) this pull request initial intention was to remove any obstacle in doing M4 internal release. It does NOT mean I closed my ears for receiving more scrutinizing remarks :)

Unresolved issues in this pull requests was all Mike's:
- potential factoring Thread.sleep() into "lower" abstraction levels in DefaultWalker -- worthwhile to consider
- locking/synchronized related questions -- non issue from my viewpoint
- logical inconsistencies in Fib number sequence, merely cosmetic (variable rename), the other one I'd skip (changing the next() method semantic, since other important part of Nexus (auto blocking) also relies on Fib sequence as is (even if it "looks" broken, this is how it was implemented).

cstamas commented:
... and the "adaptive controller" implementation.

mah0526 commented:
+1,  oh wait, this looks closed.  go drink :beer:
